core/primitives/src/block_header.rs

```diff
@@ -0,0 +1,481 @@
+    /// Last block that has doomslug finality
+    pub last_ds_final_block: CryptoHash,
+
+    /// All the approvals included in this block
+    pub approvals: Vec<Option<Signature>>,
+
+    /// Latest protocol version that this block producer has.
+    pub latest_protocol_version: ProtocolVersion,
+}
+
+/// The part of the block approval that is different for endorsements and skips
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq, Hash)]
+pub enum ApprovalInner {
+    Endorsement(CryptoHash),
+    Skip(BlockHeight),
+}
+
+/// Block approval by other block producers with a signature
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+pub struct Approval {
+    pub inner: ApprovalInner,
+    pub target_height: BlockHeight,
+    pub signature: Signature,
+    pub account_id: AccountId,
+}
+
+/// Block approval by other block producers.
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+pub struct ApprovalMessage {
+    pub approval: Approval,
+    pub target: AccountId,
+}
+
+impl ApprovalInner {
+    pub fn new(
+        parent_hash: &CryptoHash,
+        parent_height: BlockHeight,
+        target_height: BlockHeight,
+    ) -> Self {
+        if target_height == parent_height + 1 {
+            ApprovalInner::Endorsement(parent_hash.clone())
+        } else {
+            ApprovalInner::Skip(parent_height)
+        }
+    }
+}
+
+impl Approval {
+    pub fn new(
+        parent_hash: CryptoHash,
+        parent_height: BlockHeight,
+        target_height: BlockHeight,
+        signer: &dyn ValidatorSigner,
+    ) -> Self {
+        let inner = ApprovalInner::new(&parent_hash, parent_height, target_height);
+        let signature = signer.sign_approval(&inner, target_height);
+        Approval { inner, target_height, signature, account_id: signer.validator_id().clone() }
+    }
+
+    pub fn get_data_for_sig(inner: &ApprovalInner, target_height: BlockHeight) -> Vec<u8> {
+        [inner.try_to_vec().unwrap().as_ref(), target_height.to_be_bytes().as_ref()].concat()
+    }
+}
+
+impl ApprovalMessage {
+    pub fn new(approval: Approval, target: AccountId) -> Self {
+        ApprovalMessage { approval, target }
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, Eq, PartialEq)]
+pub struct BlockHeaderV1 {
+    pub prev_hash: CryptoHash,
+
+    /// Inner part of the block header that gets hashed, split into two parts, one that is sent
+    ///    to light clients, and the rest
+    pub inner_lite: BlockHeaderInnerLite,
+    pub inner_rest: BlockHeaderInnerRest,
+
+    /// Signature of the block producer.
+    pub signature: Signature,
+
+    /// Cached value of hash for this block.
+    #[borsh_skip]
+    pub hash: CryptoHash,
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, Eq, PartialEq)]
+#[borsh_init(init)]
+pub enum BlockHeader {
+    BlockHeaderV1(BlockHeaderV1),
```

Comment:
Update to `Box<BlockHeaderV1>` otherwise old block headers can be a burden in the future

Reply:
Any idea why?
```
error[E0277]: the trait bound `std::boxed::Box<block_header::BlockHeaderV1>: borsh::ser::BorshSerialize` is not satisfied
   --> core/primitives/src/block_header.rs:149:10
    |
149 | #[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, Eq, PartialEq)]
    |          ^^^^^^^^^^^^^^ the trait `borsh::ser::BorshSerialize` is not implemented for `std::boxed::Box<block_header::BlockHeaderV1>`
    |
    = help: the following implementations were found:
              <std::boxed::Box<[u8]> as borsh::ser::BorshSerialize>
    = help: see issue #48214
    = help: add `#![feature(trivial_bounds)]` to the crate attributes to enable
    = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

Reply:
Seems we didn't implement `Box<T>` 🤦

Reply:
I'll add it https://github.com/near/borsh/issues/19

Reply:
@evgenykuzyakov your concern is that the header can become smaller in the future and the old headers will dominate the size in memory?

Reply:
@evgenykuzyakov and @ilblackdragon `Box<T>` is implemented, see: https://github.com/near/borsh/blob/master/borsh-rs/borsh/src/ser/mod.rs#L278

What you are observing is most likely a conflict of several borsh versions, or you are using an outdated borsh version.

Reply:
It's in non-released version apparently. So will depend here on `master`.